### PR TITLE
Add structured logging (ECS and log level support)

### PR DESCRIPTION
**Pull request by jlvoiseux**
### Motivation / Summary
This pull request aims to used structured logging to reduce the default amount of logged statements, while providing granularity for users to adjust the verbosity of the extension. The logger used is [logrus](https://github.com/sirupsen/logrus), with an ECS formatter. Ultimately, this PR should allow the closure of #80 and #60.

### Changes
- All `log.Print` statements have been replaced by `logrus` calls and been assigned a log level
- The `ELASTIC_APM_LOG_LEVEL` environment variable is now processed
- The logger is initialized in each test

### CODE REVIEW

**Comment by trentm** [comment]
> * The `ELASTIC_LOG_LEVEL` environment variable is now processed

`ELASTIC_APM_LOG_LEVEL` ?

**Comment by jlvoiseux** [comment]
[Indeed](https://github.com/jlvoiseux/apm-aws-lambda/blob/8913d72d96c22d5849b793bdbbfc856939d7e08f/apm-lambda-extension/extension/process_env.go#L86-L90). Typo fixed in my initial comment.

File: apm-lambda-extension/main_test.go [hunk 1]
```diff
@@ -330,14 +331,14 @@ func TestAPMServerHangs(t *testing.T) {
 	start := time.Now()
 	assert.NotPanics(t, main)
 	assert.False(t, strings.Contains(apmServerLog.Data, string(Hangs)))
-	log.Printf("Success : test took %s", time.Since(start))
+	extension.Log.Printf("Success : test took %s", time.Since(start))
```
Comment (estolfo) [inline apm-lambda-extension/main_test.go:334]: Should this be info level?

**Review by estolfo — CHANGES_REQUESTED** [review]
A few requests for changes but overall it looks good!
A general comment: I think you can define a "before" hook for tests. I would recommend looking into using that instead of manually calling `InitLogger` in each test. I just did some quick research and I think defining a `TestMain` would accomplish this.

File: apm-lambda-extension/extension/process_env.go [hunk 2]
```diff
@@ -96,16 +104,17 @@ func ProcessEnv() *extensionConfig {
...
 	if config.dataReceiverServerPort == ":" {
 		config.dataReceiverServerPort = ":8200"
 	}
 	if config.apmServerUrl == "" {
-		log.Fatalln("please set ELASTIC_APM_LAMBDA_APM_SERVER, exiting")
+		Log.Fatalln("please set ELASTIC_APM_LAMBDA_APM_SERVER, exiting")
 	}
 	if config.apmServerSecretToken == "" && config.apmServerApiKey == "" {
-		log.Fatalln("please set ELASTIC_APM_SECRET_TOKEN or ELASTIC_APM_API_KEY, exiting")
+		Log.Fatalln("please set ELASTIC_APM_SECRET_TOKEN or ELASTIC_APM_API_KEY, exiting")
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/extension/process_env.go:117]: I think we actually do : this is what I observe on my dev Lambda when `ELASTIC_APM_SECRET_TOKEN` is empty : 

```
START RequestId: be5a2f7b-01ae-4f17-98cc-b90298660660 Version: $LATEST
2022/03/16 12:42:45 b4efe4dd-d4b0-48b4-88a1-d7b43eb53ed0
2022/03/16 12:42:45 Register response: {
	"functionName": "lambda-sample-dev-jlvoiseux",
	"functionVersion": "$LATEST",
	"handler": "lambda_function.lambda_handler"
}
2022/03/16 12:42:45 Could not read ELASTIC_APM_DATA_RECEIVER_TIMEOUT_SECONDS, defaulting to 15: strconv.Atoi: parsing "": invalid syntax
2022/03/16 12:42:45 please set ELASTIC_APM_SECRET_TOKEN or ELASTIC_APM_API_KEY, exiting
EXTENSION	Name: extension-dev-testflush	State: Registered	Events: [INVOKE,SHUTDOWN]
END RequestId: be5a2f7b-01ae-4f17-98cc-b90298660660
REPORT RequestId: be5a2f7b-01ae-4f17-98cc-b90298660660	Duration: 662.24 ms	Billed Duration: 663 ms	Memory Size: 128 MB	Max Memory Used: 11 MB	
RequestId: be5a2f7b-01ae-4f17-98cc-b90298660660 Error: exit status 1
Extension.Crash
```

File: apm-lambda-extension/extension/apm_server.go [hunk 3]
```diff
@@ -85,7 +84,7 @@ func PostToApmServer(client *http.Client, agentData AgentData, config *extension
 		return fmt.Errorf("failed to read the response body after posting to the APM server")
 	}
 
-	log.Printf("APM server response body: %v\n", string(body))
-	log.Printf("APM server response status code: %v\n", resp.StatusCode)
+	Log.Infof("APM server response body: %v\n", string(body))
+	Log.Infof("APM server response status code: %v\n", resp.StatusCode)
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/extension/apm_server.go]: Solved in [657ab8b](https://github.com/elastic/apm-aws-lambda/pull/147/commits/657ab8bd16219f5246640928e4bb65e20166ec6f)

File: apm-lambda-extension/extension/process_events.go [hunk 4]
```diff
@@ -19,28 +19,27 @@ package extension
...
 import (
 	"encoding/json"
-	"log"
 	"net/http"
 )
 
 func ProcessShutdown() {
-	log.Println("Received SHUTDOWN event")
-	log.Println("Exiting")
+	Log.Info("Received SHUTDOWN event")
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/extension/process_events.go]: Solved in [657ab8b](https://github.com/elastic/apm-aws-lambda/pull/147/commits/657ab8bd16219f5246640928e4bb65e20166ec6f)

File: apm-lambda-extension/extension/route_handlers.go [hunk 5]
```diff
@@ -102,9 +101,9 @@ func handleIntakeV2Events(agentDataChan chan AgentData) func(w http.ResponseWrit
 
 			select {
 			case agentDataChan <- agentData:
-				log.Println("Adding agent data to buffer to be sent to apm server")
+				Log.Info("Adding agent data to buffer to be sent to apm server")
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/extension/route_handlers.go]: Solved in [657ab8b](https://github.com/elastic/apm-aws-lambda/pull/147/commits/657ab8bd16219f5246640928e4bb65e20166ec6f)

File: apm-lambda-extension/logsapi/route_handlers.go [hunk 6]
```diff
@@ -29,14 +29,14 @@ func handleLogEventsRequest(out chan LogEvent) func(w http.ResponseWriter, r *ht
...
-			log.Printf("Error unmarshalling log events: %+v", err)
+			extension.Log.Errorf("Error unmarshalling log events: %+v", err)
 			w.WriteHeader(http.StatusInternalServerError)
 			return
 		}
 
 		for idx := range logEvents {
 			if logEvents[idx].Type == "" {
-				log.Printf("Error unmarshalling log event: %+v", logEvents[idx])
+				extension.Log.Errorf("Error unmarshalling log event: %+v", logEvents[idx])
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/logsapi/route_handlers.go]: Solved in [657ab8b](https://github.com/elastic/apm-aws-lambda/pull/147/commits/657ab8bd16219f5246640928e4bb65e20166ec6f)

File: apm-lambda-extension/logsapi/subscribe.go [hunk 7]
```diff
@@ -95,7 +95,7 @@ func startHTTPServer(out chan LogEvent) error {
 	}
 
 	go func() {
-		log.Printf("Extension listening for logsAPI events on %s", Listener.Addr().String())
+		extension.Log.Infof("Extension listening for logsAPI events on %s", Listener.Addr().String())
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/logsapi/subscribe.go]: Solved in [657ab8b](https://github.com/elastic/apm-aws-lambda/pull/147/commits/657ab8bd16219f5246640928e4bb65e20166ec6f)

File: apm-lambda-extension/main.go [hunk 8]
```diff
@@ -42,24 +41,28 @@ var (
...
+	extension.InitLogger()
+
 	sigs := make(chan os.Signal, 1)
 	signal.Notify(sigs, syscall.SIGTERM, syscall.SIGINT)
 	go func() {
 		s := <-sigs
 		cancel()
-		log.Printf("Received %v\n", s)
-		log.Println("Exiting")
+		extension.Log.Infof("Received %v\n", s)
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/main.go]: Solved in [657ab8b](https://github.com/elastic/apm-aws-lambda/pull/147/commits/657ab8bd16219f5246640928e4bb65e20166ec6f)

File: apm-lambda-extension/main.go [hunk 9]
```diff
@@ -42,24 +41,28 @@ var (
...
 	// register extension with AWS Extension API
 	res, err := extensionClient.Register(ctx, extensionName)
 	if err != nil {
 		panic(err)
 	}
-	log.Printf("Register response: %v\n", extension.PrettyPrint(res))
-
-	// pulls ELASTIC_ env variable into globals for easy access
-	config := extension.ProcessEnv()
+	extension.Log.Tracef("Register response: %v\n", extension.PrettyPrint(res))
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/main.go]: Solved in [657ab8b](https://github.com/elastic/apm-aws-lambda/pull/147/commits/657ab8bd16219f5246640928e4bb65e20166ec6f)

File: apm-lambda-extension/main.go [hunk 10]
```diff
@@ -97,14 +100,14 @@ func main() {
...
 			event, err := extensionClient.NextEvent(ctx)
 			if err != nil {
-				log.Printf("Error: %v\n", err)
-				log.Println("Exiting")
+				extension.Log.Errorf("Error: %v\n. Exiting.", err)
 				return
 			}
-			log.Printf("Received event: %v\n", extension.PrettyPrint(event))
+			extension.Log.Info("Received event.")
+			extension.Log.Tracef("%v\n", extension.PrettyPrint(event))
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/main.go]: Solved in [657ab8b](https://github.com/elastic/apm-aws-lambda/pull/147/commits/657ab8bd16219f5246640928e4bb65e20166ec6f)

File: apm-lambda-extension/main.go [hunk 11]
```diff
@@ -134,12 +137,12 @@ func main() {
 				for {
 					select {
 					case <-funcDone:
-						log.Println("funcDone signal received, not processing any more agent data")
+						extension.Log.Info("funcDone signal received, not processing any more agent data")
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/main.go]: Solved in [657ab8b](https://github.com/elastic/apm-aws-lambda/pull/147/commits/657ab8bd16219f5246640928e4bb65e20166ec6f)

File: apm-lambda-extension/main.go [hunk 12]
```diff
@@ -151,19 +154,19 @@ func main() {
 				for {
 					select {
 					case <-funcDone:
-						log.Println("funcDone signal received, not processing any more log events")
+						extension.Log.Info("funcDone signal received, not processing any more log events")
 						return
 					case logEvent := <-logsChannel:
-						log.Printf("Received log event %v\n", logEvent.Type)
+						extension.Log.Infof("Received log event %v\n", logEvent.Type)
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/main.go]: Solved in [657ab8b](https://github.com/elastic/apm-aws-lambda/pull/147/commits/657ab8bd16219f5246640928e4bb65e20166ec6f)

File: apm-lambda-extension/main.go [hunk 13]
```diff
@@ -180,11 +183,11 @@ func main() {
 
 			select {
 			case <-extension.AgentDoneSignal:
-				log.Println("Received agent done signal")
+				extension.Log.Info("Received agent done signal")
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/main.go]: Solved in [657ab8b](https://github.com/elastic/apm-aws-lambda/pull/147/commits/657ab8bd16219f5246640928e4bb65e20166ec6f)

**Comment by jlvoiseux** [comment]
Thanks for the review @estolfo ! I have implemented your suggested changes. For illustration purposes, please find below the output for all supported log levels. If that is suitable, I believe we can merge those changes. Note that the `ELASTIC_APM_LOG_LEVEL` also impacts the output of the agent (Python agent used here).

**Warning level**
```
START RequestId: 8ad116cf-a57d-42a8-bf02-9b8c331f8414 Version: $LATEST
{"@timestamp":"2022-03-16T13:07:53.823Z","ecs.version":"1.6.0","log.level":"warning","message":"Could not read ELASTIC_APM_DATA_RECEIVER_TIMEOUT_SECONDS, defaulting to 15: strconv.Atoi: parsing \"\": invalid syntax\n"}
{"@timestamp":"2022-03-16T13:07:53.823Z","ecs.version":"1.6.0","log.level":"warning","message":"Could not read ELASTIC_APM_DATA_FORWARDER_TIMEOUT_SECONDS, defaulting to 3: strconv.Atoi: parsing \"\": invalid syntax\n"}
LOGS	Name: extension-dev-linux-amd64	State: Subscribed	Types: [platform]
EXTENSION	Name: extension-dev-linux-amd64	State: Ready	Events: [INVOKE,SHUTDOWN]
END RequestId: 8ad116cf-a57d-42a8-bf02-9b8c331f8414
REPORT RequestId: 8ad116cf-a57d-42a8-bf02-9b8c331f8414	Duration: 173.67 ms	Billed Duration: 174 ms	Memory Size: 128 MB	Max Memory Used: 77 MB	Init Duration: 464.09 ms
```

**Info level**
```
Function Logs
START RequestId: d851ab22-336f-4979-9ae0-364d11d5eb25 Version: $LATEST
{"@timestamp":"2022-03-16T13:08:46.608Z","ecs.version":"1.6.0","log.level":"warning","message":"Could not read ELASTIC_APM_DATA_RECEIVER_TIMEOUT_SECONDS, defaulting to 15: strconv.Atoi: parsing \"\": invalid syntax\n"}
{"@timestamp":"2022-03-16T13:08:46.608Z","ecs.version":"1.6.0","log.level":"warning","message":"Could not read ELASTIC_APM_DATA_FORWARDER_TIMEOUT_SECONDS, defaulting to 3: strconv.Atoi: parsing \"\": invalid syntax\n"}
{"@timestamp":"2022-03-16T13:08:46.616Z","ecs.version":"1.6.0","log.level":"info","message":"Extension listening for apm data on :8200"}
{"@timestamp":"2022-03-16T13:08:46.630Z","ecs.version":"1.6.0","log.level":"info","message":"Extension listening for Lambda Logs API events on 169.254.79.130:45673"}
LOGS	Name: extension-dev-linux-amd64	State: Subscribed	Types: [platform]
{"@timestamp":"2022-03-16T13:08:46.633Z","ecs.version":"1.6.0","log.level":"info","message":"Waiting for next event..."}
EXTENSION	Name: extension-dev-linux-amd64	State: Ready	Events: [INVOKE,SHUTDOWN]
[INFO]	2022-03-16T13:08:47.20Z	d851ab22-336f-4979-9ae0-364d11d5eb25	Fetched APM Server version 8.1.0
{"@timestamp":"2022-03-16T13:08:47.127Z","ecs.version":"1.6.0","log.level":"info","message":"Received runtimeDone event for this function invocation"}
{"@timestamp":"2022-03-16T13:08:47.176Z","ecs.version":"1.6.0","log.level":"info","message":"Waiting for next event..."}
END RequestId: d851ab22-336f-4979-9ae0-364d11d5eb25
REPORT RequestId: d851ab22-336f-4979-9ae0-364d11d5eb25	Duration: 179.97 ms	Billed Duration: 180 ms	Memory Size: 128 MB	Max Memory Used: 77 MB	Init Duration: 432.01 ms
```

**Debug level**
```
09:41.755Z		Skipping instrumentation of httpx. Module httpx not found
[DEBUG]	2022-03-16T13:09:41.756Z		Skipping instrumentation of pymongo. Module pymongo.collection not found
[DEBUG]	2022-03-16T13:09:41.756Z		Skipping instrumentation of httpcore. Module httpcore._sync.connection not found
[INFO]	2022-03-16T13:09:41.757Z		Fetched APM Server version 8.1.0
[DEBUG]	2022-03-16T13:09:41.759Z		Skipping instrumentation of pylibmc. Module pylibmc not found
[DEBUG]	2022-03-16T13:09:41.759Z		Skipping instrumentation of aioredis. Module aioredis.commands.transaction not found
[DEBUG]	2022-03-16T13:09:41.759Z		Skipping instrumentation of redis. Module redis.connection not found
[DEBUG]	2022-03-16T13:09:41.760Z		Instrumented requests, requests.sessions.Session.send
EXTENSION	Name: extension-dev-linux-amd64	State: Ready	Events: [INVOKE,SHUTDOWN]
{"@timestamp":"2022-03-16T13:09:41.762Z","ecs.version":"1.6.0","log.level":"debug","message":"Received event."}
{"@timestamp":"2022-03-16T13:09:41.762Z","ecs.version":"1.6.0","log.level":"debug","message":"{\n\t\"eventType\": \"INVOKE\",\n\t\"deadlineMs\": 1647436201761,\n\t\"requestId\": \"5fb51773-c88c-4437-b432-c2d0bbcbb940\",\n\t\"invokedFunctionArn\": \"arn:aws:lambda:eu-central-1:627286350134:function:lambda-sample-dev-jlvoiseux\",\n\t\"tracing\": {\n\t\t\"type\": \"X-Amzn-Trace-Id\",\n\t\t\"value\": \"Root=1-6231e195-3908b10c565b697c3c7eabe0;Parent=77b366bf6ee67f3b;Sampled=0\"\n\t}\n}\n"}
{"@timestamp":"2022-03-16T13:09:41.762Z","ecs.version":"1.6.0","log.level":"debug","message":"Checking for agent data"}
{"@timestamp":"2022-03-16T13:09:41.762Z","ecs.version":"1.6.0","log.level":"debug","message":"No agent data on buffer"}
{"@timestamp":"2022-03-16T13:09:41.762Z","ecs.version":"1.6.0","log.level":"debug","message":"Received log event platform.logsSubscription\n"}
[DEBUG]	2022-03-16T13:09:41.762Z	5fb51773-c88c-4437-b432-c2d0bbcbb940	flushing elasticapm
[DEBUG]	2022-03-16T13:09:41.772Z	5fb51773-c88c-4437-b432-c2d0bbcbb940	forced flush
{"@timestamp":"2022-03-16T13:09:41.792Z","ecs.version":"1.6.0","log.level":"debug","message":"Received log event platform.start\n"}
{"@timestamp":"2022-03-16T13:09:41.792Z","ecs.version":"1.6.0","log.level":"debug","message":"Received log event platform.extension\n"}
{"@timestamp":"2022-03-16T13:09:41.793Z","ecs.version":"1.6.0","log.level":"debug","message":"Adding agent data to buffer to be sent to apm server"}
[DEBUG]	2022-03-16T13:09:41.793Z	5fb51773-c88c-4437-b432-c2d0bbcbb940	Sent request, url=http://localhost:8200/intake/v2/events size=0.67kb status=202
[DEBUG]	2022-03-16T13:09:41.813Z	5fb51773-c88c-4437-b432-c2d0bbcbb940	done flushing elasticapm
{"@timestamp":"2022-03-16T13:09:41.840Z","ecs.version":"1.6.0","log.level":"debug","message":"Received log event platform.runtimeDone\n"}
{"@timestamp":"2022-03-16T13:09:41.840Z","ecs.version":"1.6.0","log.level":"info","message":"Received runtimeDone event for this function invocation"}
{"@timestamp":"2022-03-16T13:09:41.840Z","ecs.version":"1.6.0","log.level":"debug","message":"Received runtimeDone signal"}
{"@timestamp":"2022-03-16T13:09:41.929Z","ecs.version":"1.6.0","log.level":"debug","message":"APM server response body: \n"}
{"@timestamp":"2022-03-16T13:09:41.929Z","ecs.version":"1.6.0","log.level":"debug","message":"APM server response status code: 202\n"}
{"@timestamp":"2022-03-16T13:09:41.929Z","ecs.version":"1.6.0","log.level":"debug","message":"Received signal that function has completed, not processing any more agent data"}
{"@timestamp":"2022-03-16T13:09:41.929Z","ecs.version":"1.6.0","log.level":"debug","message":"Checking for agent data"}
{"@timestamp":"2022-03-16T13:09:41.929Z","ecs.version":"1.6.0","log.level":"debug","message":"No agent data on buffer"}
{"@timestamp":"2022-03-16T13:09:41.929Z","ecs.version":"1.6.0","log.level":"info","message":"Waiting for next event..."}
END RequestId: 5fb51773-c88c-4437-b432-c2d0bbcbb940
REPORT RequestId: 5fb51773-c88c-4437-b432-c2d0bbcbb940	Duration: 168.36 ms	Billed Duration: 169 ms	Memory Size: 128 MB	Max Memory Used: 76 MB	Init Duration: 450.69 ms
```

File: apm-lambda-extension/extension/process_env.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (trentm) [reply apm-lambda-extension/extension/process_env.go:117]: My understanding is that `.Fatal*(...)` will call `os.Exit(1)` after logging, per https://github.com/sirupsen/logrus#level-logging

**Comment by trentm** [comment]
```
{"@timestamp":"2022-03-16T13:11:20.698Z","ecs.version":"1.6.0","log.level":"warning","message":"Could not read ELASTIC_APM_DATA_RECEIVER_TIMEOUT_SECONDS, defaulting to 15: strconv.Atoi: parsing \"\": invalid syntax\n"}
```

It would be nice if there was a field in those ecs-logging formatted log records that showed this is from the extension. Then when they are mixed with ecs-logging formatted logs from another source, they could be clearly told apart. E.g. the Node.js APM agent:

```
{"log.level":"debug","@timestamp":"2022-03-16T17:09:37.355Z","log":{"logger":"elastic-apm-node"},"ecs":{"version":"1.6.0"},"message":"request details: {\"protocol\":\"http:\",\"host\":\"httpstat.us\",\"id\":\"979bfd5934c179b5\"}"}
```

The Node.js APM agent is setting the `log.logger` field (XXX).
However, when we discussed briefly on a call a week or so ago, I think @felixbarny suggested using the "event.dataset" field (https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-dataset). So perhaps something like `"event.dataset":"apm-lambda-extension"`. I don't know the logrus syntax to add that.

**Comment by trentm** [comment]
> I don't know the logrus syntax to add that.

https://github.com/sirupsen/logrus#default-fields

**Comment by jlvoiseux** [comment]
Hello @trentm ! Thank you for the suggestion 😄 
I have added a default `event.dataset` to all log entries in [64ab892](https://github.com/elastic/apm-aws-lambda/pull/147/commits/64ab89244b6def44245e9750587666423d3f0586). This is what the resulting log looks like : 

```
START RequestId: 57f65514-8863-4689-a7e0-978e09d3951a Version: $LATEST
{"@timestamp":"2022-03-17T09:32:13.142Z","ecs.version":"1.6.0","event.dataset":"apm-lambda-extension","log.level":"debug","message":"Received event."}
{"@timestamp":"2022-03-17T09:32:13.142Z","ecs.version":"1.6.0","event.dataset":"apm-lambda-extension","log.level":"debug","message":"{\n\t\"eventType\": \"INVOKE\",\n\t\"deadlineMs\": 1647509553141,\n\t\"requestId\": \"57f65514-8863-4689-a7e0-978e09d3951a\",\n\t\"invokedFunctionArn\": \"arn:aws:lambda:eu-central-1:627286350134:function:lambda-sample-dev-jlvoiseux\",\n\t\"tracing\": {\n\t\t\"type\": \"X-Amzn-Trace-Id\",\n\t\t\"value\": \"Root=1-6233001d-761b411f269e1ede5aba4067;Parent=24b08f0855704196;Sampled=0\"\n\t}\n}\n"}
{"@timestamp":"2022-03-17T09:32:13.142Z","ecs.version":"1.6.0","event.dataset":"apm-lambda-extension","log.level":"debug","message":"Checking for agent data"}
{"@timestamp":"2022-03-17T09:32:13.142Z","ecs.version":"1.6.0","event.dataset":"apm-lambda-extension","log.level":"debug","message":"No agent data on buffer"}
{"@timestamp":"2022-03-17T09:32:13.142Z","ecs.version":"1.6.0","event.dataset":"apm-lambda-extension","log.level":"debug","message":"Received log event platform.end\n"}
{"@timestamp":"2022-03-17T09:32:13.142Z","ecs.version":"1.6.0","event.dataset":"apm-lambda-extension","log.level":"debug","message":"Received log event platform.report\n"}
[DEBUG]	2022-03-17T09:32:13.143Z	57f65514-8863-4689-a7e0-978e09d3951a	flushing elasticapm
{"@timestamp":"2022-03-17T09:32:13.159Z","ecs.version":"1.6.0","event.dataset":"apm-lambda-extension","log.level":"debug","message":"Adding agent data to buffer to be sent to apm server"}
[DEBUG]	2022-03-17T09:32:13.143Z	57f65514-8863-4689-a7e0-978e09d3951a	forced flush
{"@timestamp":"2022-03-17T09:32:13.181Z","ecs.version":"1.6.0","event.dataset":"apm-lambda-extension","log.level":"debug","message":"Received log event platform.start\n"}
[DEBUG]	2022-03-17T09:32:13.160Z	57f65514-8863-4689-a7e0-978e09d3951a	Sent request, url=http://localhost:8200/intake/v2/events size=0.68kb status=202
[DEBUG]	2022-03-17T09:32:13.181Z	57f65514-8863-4689-a7e0-978e09d3951a	done flushing elasticapm
{"@timestamp":"2022-03-17T09:32:13.198Z","ecs.version":"1.6.0","event.dataset":"apm-lambda-extension","log.level":"debug","message":"APM server response body: \n"}
{"@timestamp":"2022-03-17T09:32:13.198Z","ecs.version":"1.6.0","event.dataset":"apm-lambda-extension","log.level":"debug","message":"APM server response status code: 202\n"}
END RequestId: 57f65514-8863-4689-a7e0-978e09d3951a
REPORT RequestId: 57f65514-8863-4689-a7e0-978e09d3951a	Duration: 86.73 ms	Billed Duration: 87 ms	Memory Size: 128 MB	Max Memory Used: 77 MB
```

**Review by estolfo — APPROVED** [review]

**Comment by jlvoiseux** [comment]
Resolves #60 and #80.

File: apm-lambda-extension/extension/route_handlers.go [hunk 14]
```diff
@@ -102,9 +101,9 @@ func handleIntakeV2Events(agentDataChan chan AgentData) func(w http.ResponseWrit
 
 			select {
 			case agentDataChan <- agentData:
-				log.Println("Adding agent data to buffer to be sent to apm server")
+				Log.Debug("Adding agent data to buffer to be sent to apm server")
 			default:
-				log.Println("Channel full: dropping event")
+				Log.Warn("Channel full: dropping event")
```
[the remark this answers is not in the document]
Comment (jlvoiseux) [reply apm-lambda-extension/extension/route_handlers.go]: Solved in [2963d04](https://github.com/elastic/apm-aws-lambda/pull/147/commits/2963d04cefc36c7c0ff3578d706ec0408a9925bf)

**Comment by jlvoiseux** [comment]
Agreed, removing these trailing newlines is something we should do. I have opened issue #149 and PR #150 to clean those up.

**Pull request merged**